### PR TITLE
ci(cypress): add cypress dashboard

### PR DESCRIPTION
**Pull request by DlgSHi**
### Proposed behaviour

We've got approved OSS cypress plan for carbon so we need to:
- update all npm scripts;
- update `.travis.yml` file to exclude `cypress` tests from it;
- create a `cypress-parallel.yml` in `.github/workflows/` to run all regression using Github Actions CI;
- delete all `deprecated` cypress tests;
- delete all `docker` / `TeamCity` setup;
- clear up the `package.json` file `script` section;
- update the `README.md` files for `cypress` and `carbon`.
### Important
- When the PR will be created from a `carbon` fork - the GitHub Actions **will not pass** the `cypress record key` for that PR that's why the tests will run for 2,5 hours on 1 thread and **will not block other runs**.

After implementation the code the time for parallel tests for all regression test suite is approx ~10-12 min.
[dashboard link](https://dashboard.cypress.io/projects/8458bb/runs/130/specs)
Screenshot 2020-05-04 at 13 19 32

### Current behaviour

We are running all our cypress test on one thread. All regression suite took for a approx ~3h.

### Checklist

<del> - [ ] Release notes (Conventional Commits)  </del>
<del> - [ ] Unit tests </del>
- [x] Cypress automation tests

<del> - [ ] Storybook added or updated </del>
<del> - [ ] Theme support </del>
<del> - [ ] Typescript `d.ts` file added or updated </del> 

### Testing instructions

all suites run on Travis using two new script commands:
- npm run cypress:ci:smoke
- npm run cypress:ci:all

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -28,11 +55,47 @@ jobs:
         - npm run lint-ts
     - name: "unit tests"
       script: npm test
-    - name: "test storybook with cypress"
-      script: 
-        - npm start -- --ci </dev/null &>/dev/null &
-        - wait-on http://localhost:9001
-        - npm run test-cypress-build
+    - script:
+      name: "Cypress Smoke Tests 1st thread - chrome"
```
Comment (KatarzynaQA) [inline .travis.yml:59]:
```suggestion
      name: "Cypress Smoke Tests  on Travis CI 1st thread - chrome"
```

File: .travis.yml [hunk 2]
```diff
@@ -28,11 +55,47 @@ jobs:
...
+      env:
+      - STAGE_NAME=3x-chrome
+      <<: *cypress-travis-smoke
+    - script:
+      name: "Cypress Smoke Tests 2nd thread - chrome"
+      env:
+      - STAGE_NAME=3x-chrome
+      <<: *cypress-travis-smoke
+    - script:
+      name: "Cypress Smoke Tests 3rd thread - chrome"
```
Comment (KatarzynaQA) [inline .travis.yml:69]:
```suggestion
      name: "Cypress Smoke Tests  on Travis CI 3rd thread - chrome"
```

File: .travis.yml [hunk 3]
```diff
@@ -28,11 +55,47 @@ jobs:
...
       script: npm test
-    - name: "test storybook with cypress"
-      script: 
-        - npm start -- --ci </dev/null &>/dev/null &
-        - wait-on http://localhost:9001
-        - npm run test-cypress-build
+    - script:
+      name: "Cypress Smoke Tests 1st thread - chrome"
+      env:
+      - STAGE_NAME=3x-chrome
```
Comment (KatarzynaQA) [inline .travis.yml:61]:
```suggestion
      - STAGE_NAME=smoke-chrome
```

File: .travis.yml [hunk 4]
```diff
@@ -28,11 +55,47 @@ jobs:
...
+      <<: *cypress-travis-smoke
+    - script:
+      name: "Cypress Smoke Tests 2nd thread - chrome"
+      env:
+      - STAGE_NAME=3x-chrome
```
Comment (KatarzynaQA) [inline .travis.yml:66]:
```suggestion
      - STAGE_NAME=smoke-chrome
```

File: .travis.yml [hunk 5]
```diff
@@ -28,11 +55,47 @@ jobs:
...
+      <<: *cypress-travis-smoke
+    - script:
+      name: "Cypress Smoke Tests 3rd thread - chrome"
+      env:
+      - STAGE_NAME=3x-chrome
```
Comment (KatarzynaQA) [inline .travis.yml:71]:
```suggestion
      - STAGE_NAME=smoke-chrome
```

File: .travis.yml (hunk 1 above)
Comment (DlgSHi) [reply .travis.yml]: updated

File: .travis.yml (hunk 2 above)
Comment (DlgSHi) [reply .travis.yml]: updated

File: .travis.yml (hunk 3 above)
Comment (DlgSHi) [reply .travis.yml]: updated

File: .travis.yml (hunk 4 above)
Comment (DlgSHi) [reply .travis.yml]: updated

File: .travis.yml (hunk 5 above)
Comment (DlgSHi) [reply .travis.yml]: updated

**Comment by jamime** [comment]
Converted to draft to prevent merging while @DlgSHi works on Code Ship.

**Comment by jamime** [comment]
What is the behaviour for PR's from forked repos, will the tests still run?
Is it safe to pass `secrets.GITHUB_TOKEN` to builds on third party repos?

**Comment by DlgSHi** [comment]
> * What is the behaviour for PR's from forked repos, will the tests still run?
> * Is it safe to pass `secrets.GITHUB_TOKEN` to builds on third party repos?

@jamime thanks for your questions, 

1. It is possible to run Actions on fork - as explained on [GitHub](https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories), but Actions needs to be enabled for a fork.
2. Yes it is, for every fork repo access for a secrets is only state on `read` - as explained [here](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token)

File: cypress/support/step-definitions/common-steps.js [hunk 6]
```diff
@@ -483,3 +457,19 @@ When('I click onto root in Test directory in iFrame', () => {
 Then('totalRecords is set to {string} {word}', (totalRecords, element) => {
   pagerSummary().invoke('text').should('contain', `${totalRecords} ${element}`);
 });
+
+Then('{string} component is visible', (component) => {
+  getComponentNoIframe(component).should('exist');
```
Comment (KatarzynaQA) [inline cypress/support/step-definitions/common-steps.js:462]: should be `visible` or `exist` ? Please change the name of step or assertion

File: cypress/support/step-definitions/common-steps.js [hunk 7]
```diff
@@ -483,3 +457,19 @@ When('I click onto root in Test directory in iFrame', () => {
...
+});
+
+Then('{string} element is visible', (element) => {
+  getElementNoIframe(element).should('exist');
```
Comment (KatarzynaQA) [inline cypress/support/step-definitions/common-steps.js:466]: should be `visible` or `exist` ? Please change the name of step or assertion

File: cypress/support/step-definitions/common-steps.js [hunk 8]
```diff
@@ -483,3 +457,19 @@ When('I click onto root in Test directory in iFrame', () => {
...
+});
+
+Then('{string} element is visible by name', (element) => {
+  getElementNoIframeByName(element).should('exist');
```
Comment (KatarzynaQA) [inline cypress/support/step-definitions/common-steps.js:470]: should be `visible` or `exist` ? Please change the name of step or assertion

**Comment by jamime** [comment]
> > * What is the behaviour for PR's from forked repos, will the tests still run?
> > * Is it safe to pass `secrets.GITHUB_TOKEN` to builds on third party repos?
> 
> @jamime thanks for your questions,
> 
> 1. Yes, it is possible to run Actions on fork - as explained on [GitHub](https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories), but:
> 
> * Actions needs to be enabled for a fork.
> * `cypress record key` will not be passed to forked repo and regression will take as it took previously 2,5h and **will not be** `recorded` / `run in parallel`.
> 
> 1. Yes it is, for every fork repo access for a secrets is only state on `read` - as explained [here](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token)

Ah yes, in that link I see that

> When you create a pull request from a forked repository to the base repository, GitHub sends the pull_request event to the base repository and no pull request events occur on the forked repository.

So the fork won't run the build, but there should be an action run from our repo which is what we want 👍 

Happy with the permissions too.

File: package.json [hunk 9]
```diff
@@ -154,7 +140,6 @@
     "lodash": "^4.17.11",
     "marked": "^0.7.0",
     "mocha": "^7.1.1",
```
[the remark this answers is not in the document]
Comment (DlgSHi) [reply package.json]: removed

File: cypress/support/step-definitions/common-steps.js (hunk 6 above)
Comment (DlgSHi) [reply cypress/support/step-definitions/common-steps.js]: deleted

File: cypress/support/step-definitions/common-steps.js (hunk 7 above)
Comment (DlgSHi) [reply cypress/support/step-definitions/common-steps.js]: deleted

File: cypress/support/step-definitions/common-steps.js (hunk 8 above)
Comment (DlgSHi) [reply cypress/support/step-definitions/common-steps.js]: deleted

File: cypress/README.md [hunk 10]
```diff
@@ -32,12 +32,9 @@
 3. Install `npm install`.
 4. Run storybook `npm start`.
 5. Open a new terminal in the same path.
-6. Run cypress `npm run test-cypress`.
-7. If you would like to run specific cypress tests in command line (headless browser for continuous integration) use: `npm run test-cypress --spec 'cypress/features/[tests-type]/[featureFileName].feature'`.
-8. We have 3 test suites:
-  * Only build suite tests, which are running on Travis after every change/commit/push in the repository. To run use:`npm run test-cypress-build`;
-  * Only accessibility suite tests, which are running on TeamCity nightly and verifying do the components have the accessibility vulnerabilities. To run use:`npm run test-cypress-accessibility`;
-  * Regression test suite runs on TeamCity nightly and makes all regression tests. To use run: `npm run test-cypress-regression`.
+6. Run cypress `npx cypress open` or `npm run test-cypress`.
```
[the remark this answers is not in the document]
Comment (DlgSHi) [reply cypress/README.md]: updated

File: .github/wokflows/cypress-parallel.yml [hunk 11]
```diff
@@ -0,0 +1,45 @@
+name: Cypress tests
+
+on: [push, pull_request]
```
Comment (jamime) [inline .github/wokflows/cypress-parallel.yml:3]:
```suggestion
on: [pull_request]
```

It looks like we can run this just on `pull_request`.

https://github.com/cypress-io/github-action/issues/134

Which supports what we found on Chris's repo.

Comment (DlgSHi) [reply .github/wokflows/cypress-parallel.yml]: updated

**Comment by chrisbarber86** [comment]
@jamime If someone does PR from a fork - you can push their branch/commit to github.com/sage/carbon as a manual step and that will run CI on it our side but will update the PR status to reflect it as the commit hash is the same

**Comment by jamime** [comment]
@DlgSHi @chrisbarber86 

> When the PR will be created from a carbon fork - the GitHub Actions will not pass the cypress record key for that PR that's why the tests will run for 2,5 hours (blocking other runs by putting all other PR in queue

If we already know that PR's from forks are going to run in one thread, can we make it so we don't spawn as many runners. Maybe detect if the record key has been set, I don't think it's possible to detect a fork pr event.

**Comment by jamime** [comment]
I raised some issues

- https://github.com/cypress-io/github-action/issues/48#issuecomment-625389404
- https://github.community/t5/GitHub-Actions/How-to-detect-a-pull-request-from-a-fork/m-p/57393#M10189

We have been able to make it run on 1 runner on a fork PR and 20 runners on a regular PR. We still have a little inefficiency when 19 runners might be waiting for the last one to complete, but we can't fix this at this time.

**Pull request merged**